includes/options/class-amp-options-rest-controller.php

```diff
@@ -0,0 +1,122 @@
+	}
+
+	/**
+	 * Registers all routes for the controller.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_SERVER::READABLE,
+					'callback'            => [ $this, 'get_items' ],
+					'args'                => [],
+					'permission_callback' => [ $this, 'get_items_permissions_check' ],
+				],
+				[
+					'methods'             => WP_REST_Server::EDITABLE,
+					'callback'            => [ $this, 'update_items' ],
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::EDITABLE ),
+					'permission_callback' => [ $this, 'get_items_permissions_check' ],
+				],
+				'schema' => $this->get_public_item_schema(),
+			]
+		);
+	}
+
+	/**
+	 * Checks whether the current user has permission to retrieve options.
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( ! current_user_can( 'manage_options' ) ) {
+			return new WP_Error(
+				'amp_rest_cannot_view',
+				__( 'Sorry, you are not allowed to manage options for the AMP plugin for WordPress.', 'amp' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves all AMP plugin options.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$options    = AMP_Options_Manager::get_options();
+		$properties = $this->get_item_schema()['properties'];
+
+		$options = wp_array_slice_assoc( $options, array_keys( $properties ) );
+		return rest_ensure_response( $options );
+	}
+
+	/**
+	 * Updates AMP plugin options.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return array|WP_Error Array on success, or error object on failure.
+	 */
+	public function update_items( $request ) {
+		$params = $request->get_params();
+
+		AMP_Options_Manager::update_options( $params, array_keys( $this->get_item_schema()['properties'] ) );
+
+		return rest_ensure_response( $this->get_items( $request ) );
+	}
+
+	/**
+	 * Retrieves the schema for plugin options provided by the endpoint.
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		if ( ! $this->schema ) {
+			$this->schema = [
+				'$schema'    => 'http://json-schema.org/draft-04/schema#',
+				'title'      => 'amp-wp-options',
+				'type'       => 'object',
+				// Validation and sanitization occur in AMP_Options_Manager.
+				'properties' => [
+					Option::THEME_SUPPORT => [
+						'type' => 'string',
+					],
+					Option::READER_THEME  => [
+						'type' => 'string',
```

Comment:
This could possibly also be an enum, with all the possible reader themes being the values. Or that could be checked in `update_items`.

Reply:
Good idea. Here's the result of adding the `enum` to both schema properties in 13649dc9158565f0ba44613dc4a2d33527ff71e6:

<img width="272" alt="Screen Shot 2020-06-09 at 11 02 40 PM" src="https://user-images.githubusercontent.com/9020968/84225821-bf0f5400-aaa5-11ea-9ae6-834f67bc5fae.png">

Reply:
I've changed my mind about this, as it can result in leakage of information for user's installed themes. So I think it may be actually safer to use a `validate_callback` that does:

```php
'validate_callback' => function ( $value ) {
	// Note: The validate_callback is used instead of enum in order to prevent leaking the list of themes.
	return in_array( $value, wp_list_pluck( $this->reader_themes->get_themes(), 'slug' ), true );
},
```

Note that I've similarly required `manage_options` to get the list of reader themes in add9c38e715225deb9e7a4594edaf3ba3340d32.